components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/OAuthAppTenantResolverValve.java

```diff
@@ -0,0 +1,142 @@
+                if (StringUtils.isEmpty(clientId)) {
+                    try {
+                        String[] credentials = OAuth2Util.extractCredentialsFromAuthzHeader(request);
+                        if (credentials.length == 2 && StringUtils.isNotEmpty(credentials[0])) {
+                            clientId = credentials[0];
+                        }
+                    } catch (OAuthClientAuthnException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while extracting credentials from authorization header.", e);
+                        }
+                    }
+                }
+
+                if (StringUtils.isNotEmpty(clientId)) {
+                    try {
+                        OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientIdOnly(clientId);
+                        appTenant = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+                    } catch (IdentityOAuth2Exception e) {
+                        log.error("Error while getting oauth app for client Id: " + clientId, e);
+                    } catch (InvalidOAuthClientException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while getting oauth app for client Id: " + clientId, e);
+                        }
+                    }
+                }
+
+                // Set the tenant name to thread local properties.
+                if (StringUtils.isNotEmpty(appTenant)) {
+                    IdentityUtil.threadLocalProperties.get().put(TENANT_NAME_FROM_CONTEXT, appTenant);
+                }
+            }
+            getNext().invoke(request, response);
+        } finally {
+            // Clear thread local tenant name.
+            unsetThreadLocalContextTenantName(request);
+        }
+    }
+
+    /**
+     * Check whether the request is an OAuth request.
+     *
+     * @param request Http servlet request.
+     * @return True if the request is an OAuth request.
+     */
+    private boolean isOAuthRequest(Request request) {
+
+        String requestUri = request.getRequestURI();
+        return StringUtils.isNotEmpty(requestUri) && (requestUri.startsWith("/oauth/") ||
+                requestUri.startsWith("/oauth2/"));
+    }
+
+    /**
+     * Check whether the request is an OAuth 1.0 request.
+     *
+     * @param request Http servlet request.
+     * @return True if the request is an OAuth 1.0 request.
+     */
+    private boolean isOAuth10ARequest(Request request) {
+
+        String requestUri = request.getRequestURI();
+        return StringUtils.isNotEmpty(requestUri) && requestUri.startsWith("/oauth/");
+    }
+
+    /**
+     * Unset the context tenant name from thread local properties.
+     * @param request Http servlet request.
+     */
+    private void unsetThreadLocalContextTenantName(Request request) {
+
+        if (!IdentityTenantUtil.isTenantQualifiedUrlsEnabled() && isOAuthRequest(request) &&
```

Comment:
Can remove two additional checks and just check whether thread local is not null